### PR TITLE
Display pinned vars in the receive assertions

### LINKED ISSUE #3702: ExUnit: assert_receive/3 and assert_received/2 do not print value when pinned variables is used
ExUnit: assert_receive/3 and assert_received/2 do not print value when pinned variables is used in the "No message matching" message

Showing:

    No message matching ^fruits after 0ms

instead of showing the actual value of fruits

`$ cat test/ex_unit_no_match_message_pin_operator_test.exs`
```elixir
Code.require_file "./test_helper.exs", __DIR__

defmodule ElixirBugs.ExUnitNoMatchMessagePinOperator do
  use ExUnit.Case

  test "ExUnit.assert_receive/3: match pinned variables" do
    fruits = "Apples and Pears"
    send self, fruits
    assert_receive ^fruits
    assert_receive ^fruits
  end

  test "ExUnit.asssert_received/2: match pinned variables" do
    fruits = "Apples and Pears"
    send self, fruits
    assert_received ^fruits
    assert_received ^fruits
  end
end
```

`$ mix test`
```sh
  1) test assert_receive: match pinned variables (ElixirBugs.ExUnitNoMatchMessagePinOperator)
     test/ex_unit_no_match_message_pin_operator_test.exs:13
     No message matching ^fruits after 100ms. The process mailbox is empty.
     stacktrace:
       test/ex_unit_no_match_message_pin_operator_test.exs:17



  2) test asssert_received: match pinned variables (ElixirBugs.ExUnitNoMatchMessagePinOperator)
     test/ex_unit_no_match_message_pin_operator_test.exs:6
     No message matching ^fruits after 0ms. The process mailbox is empty.
     stacktrace:
       test/ex_unit_no_match_message_pin_operator_test.exs:10
```

Code available: https://github.com/eksperimental/elixir_bugs/tree/ex_unit_no_match_message_pin_operator

### PR DESCRIPTION
**Pull request by lexmag**
Closes #3702.

### CODE REVIEW

**Comment by josevalim** [comment]
It looks great so far!

**Comment by lexmag** [comment]
This is ready for review.
Last commit introduces the following formatting:
```
  1) test assert receive with pinned variable (ExUnit.AssertionsTest)
     lib/ex_unit/test/ex_unit/assertions_test.exs:110
     No message matching {:status, ^status} after 100ms.
     The following variables were pinned:
       status = :valid
     Process mailbox:
       {:status, :invalid}
     stacktrace:
       lib/ex_unit/test/ex_unit/assertions_test.exs:114
```

File: lib/ex_unit/lib/ex_unit/assertions.ex [hunk 1]
```diff
@@ -282,6 +282,13 @@ defmodule ExUnit.Assertions do
 
   defp do_assert_receive(expected, timeout, message) do
     binary = Macro.to_string(expected)
+    {_, pins} =
```
Comment (josevalim) [inline lib/ex_unit/lib/ex_unit/assertions.ex]: Maybe move this to a function? `collect_pins/1`?

**Comment by josevalim** [comment]
Just one tiny comment, everything else is excellent, thank you!

**Pull request merged**